### DISCUSSION TITLE
Dedicated automated testing page

### DISCUSSION THREAD

**Discussion by andy-hook** [radix-ui/website#810] [Documentation]
It would be helpful to provide some type of documentation regarding automated testing and primitives, especially with suites running `testing-library` and `user-event`. There are quite a few gotchas, especially common are pointer event firing problems and mocking layout specific browser apis as `jsdom` struggles with these.

Examples of common questions:

https://github.com/radix-ui/primitives/issues/1382
https://github.com/radix-ui/primitives/issues/1220
https://github.com/radix-ui/primitives/issues/1207
https://github.com/radix-ui/primitives/issues/1036
https://github.com/radix-ui/primitives/issues/856
https://github.com/radix-ui/primitives/issues/765
https://github.com/radix-ui/primitives/issues/420

Document null during testing
https://github.com/radix-ui/primitives/issues/1707

**Comment by Kikketer** [comment]
Slider continues to have an issue using vitest even with the mock setup:
```
/**
 * JSDOM doesn't implement PointerEvent so we need to mock our own implementation
 * Default to mouse left click interaction
 * https://github.com/radix-ui/primitives/issues/1822
 * https://github.com/jsdom/jsdom/pull/2666
 */
class MockPointerEvent extends Event {
  button: number;
  ctrlKey: boolean;
  pointerType: string;

  constructor(type: string, props: PointerEventInit) {
    super(type, props);
    this.button = props.button || 0;
    this.ctrlKey = props.ctrlKey || false;
    this.pointerType = props.pointerType || 'mouse';
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
window.PointerEvent = MockPointerEvent as any;
window.HTMLElement.prototype.scrollIntoView = vi.fn();
window.HTMLElement.prototype.releasePointerCapture = vi.fn();
window.HTMLElement.prototype.hasPointerCapture = vi.fn();
```

Clicking on a thumb will respond fine but still throws the 
```
Uncaught [TypeError: target.setPointerCapture is not a function]
```

Mocking it:
```
window.HTMLElement.prototype.setPointerCapture = vi.fn();
```
in that setup prevents any callbacks.
